### PR TITLE
Serve verified uncommitted blocks to seq-only lookups in CachedStorage

### PR DESCRIPTION
**Pull request by samliok**
CachedStorage keeps verified-but-not-yet-finalized blocks in a cache keyed by digest. The msm sealing-block checks (`buildBlockOrTransitionEpoch`, `areWeReadyToTransitionEpoch`) look blocks up by sequence only and pass a zero digest, so they can never hit the cache. The lookup falls through to committed storage, and while the sealing block is notarized but not yet indexed, block building fails with `failed to retrieve sealing block for previous epoch`. The failed build is not retried, so the leader misses its round and the round empty-notarizes.

### CODE REVIEW

File: adapters.go [hunk 1]
```diff
@@ -114,6 +114,16 @@ func (cs *CachedStorage) RetrieveBlock(seq uint64, digest common.Digest) (metada
 func (cs *CachedStorage) Retrieve(seq uint64, digest common.Digest) (common.VerifiedBlock, *common.Finalization, error) {
 	cs.lock.RLock()
 	item, exists := cs.cache[digest]
+	if !exists && digest == (common.Digest{}) {
+		// Seq-only lookups pass a zero digest, so scan the cache by seq.
+		// Otherwise a verified but not yet finalized block is invisible to them.
+		for _, cb := range cs.cache {
```
Comment (yacovm) [inline adapters.go]: can we iterate with index instead of this?

Comment (yacovm) [reply adapters.go]:
```
	if !exists && digest == (common.Digest{}) {
		// Seq-only lookups pass a zero digest, so scan the cache by seq.
		// Otherwise a verified but not yet finalized block is invisible to them.
		for i := range cs.cache {
			cb := cs.cache[i].ParsedBlock
			md := &cb.Metadata
			protocolMD := &md.SimplexProtocolMetadata
			if protocolMD.Seq == seq {
				item, exists = cs.cache[i], true
				break
			}
		}
	}
```

Comment (samliok) [reply adapters.go]: updated by claude 76738bc

File: adapters.go [hunk 2]
```diff
@@ -114,6 +114,20 @@ func (cs *CachedStorage) RetrieveBlock(seq uint64, digest common.Digest) (metada
 func (cs *CachedStorage) Retrieve(seq uint64, digest common.Digest) (common.VerifiedBlock, *common.Finalization, error) {
 	cs.lock.RLock()
 	item, exists := cs.cache[digest]
+	if !exists && digest == (common.Digest{}) {
```
Comment (yacovm) [inline adapters.go]: Let's add a test that tests the added lines.

Comment (samliok) [inline adapters.go]: this should be 

	if !exists || digest == (common.Digest{}) {

Comment (samliok) [reply adapters.go]: updated by claude abf52e5

Comment (samliok) [reply adapters.go]: updated by claude c9de1c6

File: adapters.go [hunk 3]
```diff
@@ -113,22 +113,34 @@ func (cs *CachedStorage) RetrieveBlock(seq uint64, digest common.Digest) (metada
 
 func (cs *CachedStorage) Retrieve(seq uint64, digest common.Digest) (common.VerifiedBlock, *common.Finalization, error) {
 	cs.lock.RLock()
+	defer cs.lock.RUnlock()
+
 	item, exists := cs.cache[digest]
 	if exists {
-		cs.lock.RUnlock()
```
Comment (yacovm) [inline adapters.go:120]: why the locking change?

File: adapters.go [hunk 4]
```diff
@@ -113,22 +113,34 @@ func (cs *CachedStorage) RetrieveBlock(seq uint64, digest common.Digest) (metada
...
 	// We don't populate the cache here because we populate it externally.
-
 	block, finalization, err := cs.GetBlock(seq)
 	if err != nil {
 		return nil, nil, err
 	}
 
+	// If we are requesting an empty digest, and the sequence is not found in storage,
+	// check if we have verified but not indexed this sequence
+	if err == common.ErrBlockNotFound && digest == (common.Digest{}) {
```
Comment (yacovm) [inline adapters.go]: shouldn't this be: `f err == common.ErrBlockNotFound || digest == (common.Digest{})` ?

Comment (samliok) [reply adapters.go]: no in this case it should be &&.

because || would mean we grab and return from the cache even if we returned a block & finalization above. this shouldn't happen because if we finalize a block, it shouldn't be in the cache but still ranging through the cache seems unnecessary.

Comment (yacovm) [reply adapters.go]: After reading the code again - I just noticed that we do the check after we check the storage with `cs.GetBlock(seq)` in line 126. 

Why do we do that? Shouldn't we iterate the cache *before* we ask the storage? 

What's the point of asking the storage first?

Comment (yacovm) [reply adapters.go]: Can't we do this?


```
func (cs *CachedStorage) Retrieve(seq uint64, digest common.Digest) (common.VerifiedBlock, *common.Finalization, error) {
	cs.lock.RLock()
	defer cs.lock.RUnlock()

	// A cached block is not finalized yet, since indexing removes it from the cache.
	if item, exists := cs.cache[digest]; exists {
		return item.ParsedBlock, nil, nil
	}

	for _, cb := range cs.cache {
		if cb.BlockHeader().Seq == seq && digest == (common.Digest{}) {
			return cb.ParsedBlock, nil, nil
		}
	}

	// We don't populate the cache here because we populate it externally.
	block, finalization, err := cs.GetBlock(seq)
	if err == nil {
		return &ParsedBlock{
			StateMachineBlock: block,
			msm:               cs.msm,
		}, finalization, nil
	}

	return nil, nil, err
}
```

Comment (samliok) [reply adapters.go]: yep, i updated the logic + re-added the locking stuff from before

File: adapters.go [hunk 5]
```diff
@@ -113,26 +113,33 @@ func (cs *CachedStorage) RetrieveBlock(seq uint64, digest common.Digest) (metada
 
 func (cs *CachedStorage) Retrieve(seq uint64, digest common.Digest) (common.VerifiedBlock, *common.Finalization, error) {
 	cs.lock.RLock()
+
+	// A cached block is not finalized yet, since indexing removes it from the cache.
```
Comment (yacovm) [inline adapters.go]: I am not sure this comment is beneficial. 

Yes, indexing removes the cached block from the cache, but how is that relevant to retrieval?

File: adapters.go [hunk 6]
```diff
@@ -113,26 +113,33 @@ func (cs *CachedStorage) RetrieveBlock(seq uint64, digest common.Digest) (metada
...
 	if exists {
 		cs.lock.RUnlock()
-		// If the block is cached, it means it's not finalized yet, because upon finalizing the block (indexing)
-		// we also remove it from the cache. Therefore, we return nil for the finalization.
 		return item.ParsedBlock, nil, nil
 	}
-	cs.lock.RUnlock()
 
-	// We don't populate the cache here because we populate it externally.
+	if digest != (common.Digest{}) {
```
Comment (yacovm) [inline adapters.go]: why do we return here?

File: adapters.go [hunk 7]
```diff
@@ -113,26 +113,33 @@ func (cs *CachedStorage) RetrieveBlock(seq uint64, digest common.Digest) (metada
...
+		cs.lock.RUnlock()
+		return nil, nil, common.ErrBlockNotFound
+	}
 
-	block, finalization, err := cs.GetBlock(seq)
-	if err != nil {
-		return nil, nil, err
+	for _, cb := range cs.cache {
+		if cb.BlockHeader().Seq == seq {
```
Comment (yacovm) [inline adapters.go:127]: we should only return if the digest either matches or the digest we requested to match is empty.

File: adapters_test.go [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+				Payload: []byte(payload),
+			},
+		},
+	}
+}
+
+// TestCachedStorageRetrieveBySeq asserts that Retrieve serves a verified but
+// not yet indexed block when the caller passes a zero digest or a digest that
+// misses the cache, by matching on seq alone.
+func TestCachedStorageRetrieveBySeq(t *testing.T) {
```
Comment (yacovm) [inline adapters_test.go]: I think we should add tests that cover the corner cases I pointed out above

File: adapters.go (hunk 5 above)
Comment (samliok) [reply adapters.go]: i guess its relevant because we don't return a finalization if we get a cache hit. i can remove the comment though

File: adapters.go (hunk 7 above)
Comment (samliok) [reply adapters.go:127]: fixed, also fixed after cs.GetBlock() is returned

File: adapters.go (hunk 6 above)
Comment (samliok) [reply adapters.go]: removed

File: adapters_test.go (hunk 8 above)
Comment (samliok) [reply adapters_test.go]: add some nice table tests

**Review by yacovm — APPROVED** [review]
